libs/core/execution/include/hpx/execution/detail/async_launch_policy_dispatch.hpp

```diff
@@ -83,6 +84,17 @@ namespace hpx { namespace detail {
                 std::forward<F>(f), std::forward<Ts>(ts)...));
             if (hpx::detail::has_async_policy(policy))
             {
+                // if one of the arguments is a future we play it conservatively
+                if constexpr (hpx::traits::is_future_any_v<std::decay_t<Ts>...>)
```

Comment:
Are there situations where this makes a difference? A future in `Ts` at this point should act as any type of argument, right, i.e. have no special handling? But I guess that's not the case if we need this here...

Reply:
This is a tricky one. Generally, futures that are used as arguments to other threads can lead to deadlocks if the thread that calls async is the one that will make the future ready (assuming the new thread happens to be executed directly). For this reason I made the decision to explicitly inhibit direct execution for threads that take futures as their arguments.

Reply:
Hmm, do you have a small example to demonstrate this? Not necessarily compiling code, just the gist of it.

Reply:
```
promise<void> p;
auto f = async([](future<void> f2){ f2.get(); }, p.get_future());
f.then(...);    // <-- might cause lambda to be executed directly
p.set_value();
``` 
If the lambda is executed directly, `f2.get()` will block indefinitely as the promise `p` will never be triggered.

Reply:
> I don't have ideas to avoid this though...

The only thing we can do is to be aware of this. A sudden deadlock is a plain giveaway for this kind of situation. It might be difficult to identify it in real code, however.

Reply:
> A sudden deadlock is a plain giveaway for this kind of situation.

It's plain giveaway for something being broken somewhere... But on the whole I agree with your comment.